### PR TITLE
AI SMART FIRE WORKS. WITH TESTS.

### PR DESCRIPTION
**Pull request by daniel-schroeder-io**
AI smart fire is a separate function that should be called from main

smart fire uses the original fire function if it has not yet hit a ship. 

Ive added some methods to help with refactoring. 

Tests give pretty good line coverage (93-96%) but can be coded to get higher.

### CODE REVIEW

**Comment by srb76** [comment]
I'll leave this pull request for someone who still needs to do a code review, but you may want to add some extra comments to your tests. Just say what you are testing for when you have a series of asserts without comments.

File: src/Test/BattleshipModelTest.java [hunk 1]
```diff
@@ -183,6 +183,309 @@ public void shootAtPlayer() throws Exception {
 
     }
 
+
```
Comment (tnoelcke) [inline src/Test/BattleshipModelTest.java:186]: Please add some comments in this test so we know how it works.

Thanks!

File: src/Test/BattleshipModelTest.java [hunk 2]
```diff
@@ -183,6 +183,309 @@ public void shootAtPlayer() throws Exception {
...
+        assertEquals(3, model.getHitArraySize());
+        assertEquals(0, model.getMissArraySize());
+        assertEquals(null, model.firstFireHit);
+        model.shootAtPlayer(5,1);
+        assertEquals(4, model.getHitArraySize());
+        assertEquals(0, model.getMissArraySize());
+        assertEquals(null, model.firstFireHit);
+
+    }
+
```
Comment (tnoelcke) [inline src/Test/BattleshipModelTest.java:246]: Again please add comments as these tests look rather complicated if i broke your test i might have a hard time figuring out why.

File: src/main/java/edu/oregonstate/cs361/battleship/BattleshipModel.java [hunk 3]
```diff
@@ -386,84 +399,188 @@ private Coordinate getRandomCoordinate(){
...
 
-                            if(addpoints[j].equals(coor) == false){
-                                playerHits.add(addpoints[j]);
-                            }
+        if(civilianHit == null && miliatryHit == null)
+            playerMisses.add(coor[0]);
 
-                        }
-                    }
+    }
```
Comment (tnoelcke) [inline src/main/java/edu/oregonstate/cs361/battleship/BattleshipModel.java:506]: This is super clean!

File: src/main/java/edu/oregonstate/cs361/battleship/BattleshipModel.java [hunk 4]
```diff
@@ -511,12 +628,101 @@ private Ship getShipFromCoordinate(Coordinate c){
...
+        }
+        Coordinate [] myarray = new Coordinate[length];
+        for(int i = 0; i < length; i++){
+            myarray[i] = toReturn.get(i);
+        }
+
+        return myarray;
+    }
+
+
```
Comment (tnoelcke) [inline src/main/java/edu/oregonstate/cs361/battleship/BattleshipModel.java:678]: Good thinking i like what you did here.

File: src/main/java/edu/oregonstate/cs361/battleship/BattleshipModel.java [hunk 5]
```diff
@@ -386,84 +399,188 @@ private Coordinate getRandomCoordinate(){
         return new Coordinate(randRow,randCol);
 
     }
-
-
-    public void shootAtPlayer() {
+    public void smartShootAtPlayer() {
```
Comment (tnoelcke) [inline src/main/java/edu/oregonstate/cs361/battleship/BattleshipModel.java:402]: We can talk about it tomorrow but, this function is huge. I realize its the most complicated piece of code in our program but maybe we should find a way to break it up? Lets look through this thing tomorrow and figure it out.

**Review by tnoelcke — COMMENTED** [review]
I've added a few comments about things we will probably need to change.

**Comment by tnoelcke** [comment]
People are waiting for this code. We will make the changes later and pull them in.

**Pull request merged**